### PR TITLE
add latex renderer

### LINKED ISSUE #66: [Proposal] Latex Renderer
# Summary

Render Latex Math Equations. 

# Example search queries
- Math Latex renderer
- Render Latex Math
- Latex math compile

# Context

Reference - https://www.codecogs.com/latex/eqneditor.php

### PR DESCRIPTION
**Pull request by SuyashSonawane**
This PR is related to proposal in issue #66

Closes #66 

Latex renderer is added as a InstantApp which responds to the query which contains `latex` term
>Below is output of the same.
image

The download feature is still experimental and can be improved to support full document pdf conversion
>Below is output of the resultant pdf 
image

### CODE REVIEW

File: src/apps/LatexRender.tsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        ).then(canvas => {
+            let pdf = new jsPDF();
+            pdf.addImage(canvas.toDataURL('image/png'), 'PNG', 0, 10, 215, 29);
+            pdf.save(filename);
+        });
+    }
+
+    return (
+        <Container>
+            <h2>Enter your latex string below to render 👇</h2>
```
Comment (sahil-shubham) [inline src/apps/LatexRender.tsx:58]:
```suggestion
            <h2>Enter LaTeX below 👇</h2>
```
We should  keep the instructions as simple as possible, "render" isn't a word an average person uses everyday, unlike us that is :sweat_smile:

File: src/apps/LatexRender.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    border:none;
+    border-bottom:1px solid black;
+
+    &:focus{
+        border:none;
+    }
+`;
+
+const Alert = styled.div`
+`;
```
Comment (OrkoHunter) [inline src/apps/LatexRender.tsx:23]:
```suggestion
```

Looks like it isn't being used right now.

**Review by OrkoHunter — APPROVED** [review]
Screen Shot 2021-09-05 at 10 04 21

Looks super elegant to be honest!

File: src/apps/LatexRender.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+            </div>
+            {content.length > 0 && <Button onClick={downloadLatex} title="This is still experimental">📃 download</Button>}
+        </Container>)
+};
+
+
+const parseLatexQuery = async ({ query }: queryToDataInput) => {
+    const triggerWord = "latex";
+    if (!query.includes(triggerWord)) {
```
Comment (OrkoHunter) [inline src/apps/LatexRender.tsx:75]: Should convert query to lowercase, to catch queries like `Latex`? 😅 
Screen Shot 2021-09-05 at 10 05 29

File: src/apps/LatexRender.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        setContent(e.target.value)
+    }
+    const downloadLatex = (e) => {
+        const filename = 'latex-renderer-felvin-search.pdf';
+        html2canvas(document.querySelector('#rendererElement'),
+            { scale: 5 }
+        ).then(canvas => {
+            let pdf = new jsPDF();
+            pdf.addImage(canvas.toDataURL('image/png'), 'PNG', 0, 10, 215, 29);
+            pdf.save(filename);
```
Comment (sahil-shubham) [inline src/apps/LatexRender.tsx:52]: I don't know much about how this `jsPDF` function works, but the text inside the downloaded PDF seems very squished to me. What do you think? 
image

File: src/apps/LatexRender.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    return (
+        <Container>
+            <h2>Enter your latex string below to render 👇</h2>
+            <InputField onChange={captureInput} placeholder="c = \pm \sqrt{a^2 + b^2}" />
+            <div
+                id="rendererElement"
+            >
+                <BlockMath
+                    errorColor={'#cc0000'}
```
Comment (sahil-shubham) [inline src/apps/LatexRender.tsx:64]: This error code colour doesn't seem to be working, I also wasn't able to read the complete error message. Is it possible to remove ellipsis or make the font size smaller? 
image

File: src/apps/LatexRender.tsx (hunk 3 above)
Comment (SuyashSonawane) [reply src/apps/LatexRender.tsx:75]: sure I'll update that

**Review by sahil-shubham — COMMENTED** [review]
Thank you for this PR, amazed by how quickly you ended up making this after you took up the issue. I left a few comments and would like to know what you think about the changes suggested.

File: src/apps/LatexRender.tsx (hunk 4 above)
Comment (SuyashSonawane) [reply src/apps/LatexRender.tsx:52]: Actually huge part of my time got consumed on this, I'm trying to convert it to pdf but, there is no direct way to convert html element with css to pdf, here I'm converting it into a image (base64 string) and then inserting it into the pdf at position.
Providing correct dimension is somewhat difficult. I'm trying to solve it.

Comment (sahil-shubham) [reply src/apps/LatexRender.tsx:52]: Thank you for going the extra effort to add this, I will look into this as well and try to help fix this.

Comment (SuyashSonawane) [reply src/apps/LatexRender.tsx:52]: Hi solved this 🥳, found a npm package that can convert html elements directly into pdf, it performs same actions under the hood

**Comment by SuyashSonawane** [comment]
I have fixed the issues and pushed the new commit.

**Comment by sahil-shubham** [comment]
Thank you for implementing the suggested changes so quickly, merging this PR. This would be published as a package on npmjs [here](https://www.npmjs.com/package/@felvin-search/apps), which later would be used on the main search [felvin.com](https://felvin.com). You would be able to see  your app there soon.

**Comment by SuyashSonawane** [comment]
Thanks @sahil-shubham and @OrkoHunter for feedback and suggestions.

**Pull request merged**